units/commit_queue.go

```diff
@@ -187,12 +197,20 @@ func (j *commitQueueJob) Run(ctx context.Context) {
 		}
 
 		grip.Info(message.Fields{
-			"source":  "commit queue",
-			"job_id":  j.ID(),
-			"item":    nextItem,
-			"message": "finished processing item",
+			"source":               "commit queue",
+			"job_id":               j.ID(),
+			"item":                 nextItem,
+			"message":              "finished processing commit queue item",
+			"processing_time_secs": time.Since(beginProcessingTime).Seconds(),
 		})
 	}
+	grip.Info(message.Fields{
+		"source":               "commit queue",
+		"job_id":               j.ID(),
+		"items":                nextItems,
+		"message":              "finished processing batch of commit queue items",
```

Comment:
If we wanted to go somewhere and answer the questions: "Would we benefit from increasing the batch size from 3 to 4 based on the current speed of our commit queue?" , where would someone go to get an answer for that? And can we document that somewhere?

Reply:
We'd theoretically always benefit right? It just may not be worth the added cost? (I might be misunderstanding what's wanted from this comment though)

Reply:
It's the "based on the current speed of our commit queue" that's key here. And yes, it would be benefit in a way that is worth the added cost. 

The main ask of the comment is: when I created this ticket my goal was to both generate that information, but then also to think of a way to make that information easily accessible without too much additional code reading or log grepping. So the question is: has that been thought about and can we document the answer?

Reply:
The formula for the worst case number of patches that can be ran for a commit queue of size `n` (if every item fails) can be quantified by this [pretty wild formula](https://docs.google.com/document/d/1n7xog26jPzM5j87ce5AnOeGwtWaGPKIieNQ2CirQwcI/edit#) from the initial design of when batches in the CQ were initially developed. So (output of the formula - `n`) is the number of patches we risk wasting by increasing the batch size. So if queue size is 10 for example, and batch size is 3, the worst case scenario is that we run 19 patches which is 9 more patches than we'd have needed to run if we didn't do any parallel execution.

Of course items don't all fail, in fact most succeed, but to track how many failures occur for each project I added some logs to `DequeueAndRestartForTask` so we can collect statistics moving forward on how many CQ items are getting dequeued due to failures over time. As this collects stats over time we can probably get a decent idea of the rate of failures to expect in each commit queue, and plug it into the above formula to get a more quantitative idea on how much work we are re-computing as a consequence of a certain batch size.

Also, I created a [Commit Queue Wait Statistics](https://mongodb.splunkcloud.com/en-US/app/search/commit_queue_wait_time_2?form.global_time.earliest=-30d%40d&form.global_time.latest=now) dashboard which currently shows project's commit queue size over time, and the # of minutes queue items wait to get processed to get better idea for the queue's performance over time. I also added a new field to the items (`QueueLengthAtEnqueue`) which saves the length of the queue at the time of enqueueing, so we can later add charts to get an accurate rate of progress of project's commit queues by comparing this value to the time it took for the items to get processed.

Reply:
Love this dashboard! I would suggest sending to product (Maria/Cris) ; they may have something similar or may wish to hold onto it.

Reply:
Sent!

Reply:
Thank you for typing that up. This information and dashboard are awesome! I'm personally a big fan of transferring this information and a link to the dashboard to one of our docs, probably the ops guide so that engineers can easily find it if they hit a commit queue conundrum. 

> I also added a new field to the items (QueueLengthAtEnqueue) which saves the length of the queue at the time of enqueueing, so we can later add charts to get an accurate rate of progress of project's commit queues by comparing this value to the time it took for the items to get processed.

I think it would be a good idea to do that now while the information is fresh, maybe in a new ticket.

Reply:
Sounds good, added it to the [commit queue troubleshooting portion](https://docs.google.com/document/d/14BTuPnzbSLCuewcMXFNQivkyUPy3Dsy1TYdF_9WVaBY/edit#heading=h.lrutrlkph14l) of the ops guide.

I think we'll need at least a few days of `QueueLengthAtEnqueue` logs being available before we make dashboards based off of it, so creating a new ticket: https://jira.mongodb.org/browse/EVG-18895